aeternity/ae_mdw
# feat: async account aex9 presence

## What

Runs AEX9 account presence sync update asynchronously.

## Why

Blocking calls during Mdw database sync are taking a **long time**.
Fixes #268 

## Problem snapshot
Times in milliseconds for each event log entry of aex9 contract during the sync.

![Screenshot from 2021-10-08 00-00-43](https://user-images.githubusercontent.com/44991200/136514432-8e04d6b7-5aa2-41e8-853f-aa88aba84199.png)

## Validation steps

Validation based on async tasks database state:
1. `make shell`
2. `require Ex2ms; require AeMdw.Db.Model; import AeMdw.Db.Util; alias AeMdw.Db.Model;`
3. `{tasks1, _cont} = select(Model.AsyncTasks, Ex2ms.fun do record -> record end, 1000)`
4. Wait 30 seconds or so...
5. `{tasks2, _cont} = select(Model.AsyncTasks, Ex2ms.fun do record -> record end, 1000)`
6. `tasks1 -- tasks2 == tasks1` # should always be true indicating it's consuming all events
7. Repeat 3-6 some times.

PS: This validation relies on the fact that the `AeMdw.Sync.AsyncTasks.Consumer.process/1` code is the same and the on aex9 presence tests found here: `test/ae_mdw/db/contract_test.exs`
